### DISCUSSION TITLE
Failing to get data from WiCAN by ESP32

### DISCUSSION THREAD

**Discussion by ashton-danci** [meatpiHQ/wican-fw#249] [General]
Hi, I have been trying to get my RPM from the meatPi via my ESP32 on my Mazda MX5 Mk2.5, but I am having no luck and constant communication errors, I am using all the default settings, can anyone identify what I have done wrong?
	Received char: I
	Received char: T
	Received char: :
	Received char: _
	Received char: .
	Received char: .
	Received char: .
	Received char: E
	Received char: R
	Received char: R
	Received char: O
	Received char: R
	Received char: \r
	Received char: \r
	Received char: >
Delimiter found.
All chars received: BUSINIT:...ERROR
ELM responded with "ERROR"
Setting protocol via AT TP A%c did not work - trying via AT SP %c
Clearing input serial buffer
Sending the following command/query: AT SP 0
	Received char: O
	Received char: K
	Received char: \r
	Received char: \n
	Received char: \r
	Received char: >
Delimiter found.
All chars received: OK
Clearing input serial buffer
Sending the following command/query: AT SP 3
Clearing input serial buffer
Sending the following command/query: AT Z
Clearing input serial buffer
Sending the following command/query: AT E0
Clearing input serial buffer
Sending the following command/query: AT L0
Clearing input serial buffer
Sending the following command/query: ATSP 0

It seems to return ERROR but then OK after it - But when I try to get my RPM it returns 0 and gives errors, here is the code: 

IPAddress server(192, 168, 80, 1);
WiFiClient client;
ELM327 myELM327;

uint32_t rpm = 0;

void setup()
{
  Serial.begin(115200);

  // Connecting to ELM327 WiFi
  Serial.print("Connecting to ");
  Serial.println(ssid);

  WiFi.mode(WIFI_AP);
  WiFi.begin(ssid, password); 

  while (WiFi.status() != WL_CONNECTED)
  {
    delay(500);
    Serial.print(".");
  }

  Serial.println("");
  Serial.println("Connected to Wifi");
  Serial.println("IP address: ");
  Serial.println(WiFi.localIP());
  Serial.print(server.toString(true));

  if (client.connect(server, 3333)){
    Serial.println("connected");
    myELM327.begin(client, true, 2000);
    myELM327.sendCommand("AT SP 3"); 
    delay(2000); 
  }
  else
  {
    Serial.println("connection failed");
    while(1);
  }
  myELM327.sendCommand("AT Z"); // Reset all 
  delay(2000); 
  myELM327.sendCommand("AT E0"); // Echo Off 
  delay(2000); 
  myELM327.sendCommand("AT L0"); // Linefeed Off 
  delay(2000); 
  myELM327.sendCommand("ATSP 0"); // Auto select protocol 
  delay(2000);

}

void loop() { 
  float tempRPM = myELM327.rpm(); 
  if (myELM327.nb_rx_state == ELM_SUCCESS) { 
    rpm = (uint32_t)tempRPM; 
    Serial.print("RPM: "); 
    Serial.println(rpm); 
    } else if (myELM327.nb_rx_state != ELM_GETTING_MSG) { 
      myELM327.printError();
      }
   }

**Comment by meatpiHQ** [comment]
@breezy-developments The issue here is that SP 0 is not supported. You have to set the protocol manually.

Also only CAN protocols are supported. Do you know the supported protocol?

**Reply by ashton-danci** [reply]
Thank you for the response its greatly appreciated.

My protocol I believe is SP 3 which I tried in a later iteration of the code

I believe my issue is that my Protocol is the K-line (ISO 9141-2) protocol, I have found all other ELM scanners work with it but they sadly do not have the flexibility the meatPi has.

Would the main issue be that I have a K-line protocol?

**Reply by meatpiHQ** [reply]
Yeah that's the problem. WiCAN only supports CAN, the PRO version supports K-line and all other protocols.

**Reply by ashton-danci** [reply]
Yikes there is my issue, thought I was going insane for a few days, Thank you for the superb help! Will get the PRO once it releases
